src/backtracking/knight_tour.rs

```diff
@@ -0,0 +1,182 @@
+    /// * `size_y` - The height of the chessboard.
+    ///
+    /// # Returns
+    ///
+    /// A new KnightTour instance.
+    fn new(size_x: usize, size_y: usize) -> Self {
+        let board = vec![vec![0; size_x]; size_y];
+        KnightTour { board }
+    }
+
+    /// Checks if the given position is safe to move to.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The x-coordinate of the position.
+    /// * `y` - The y-coordinate of the position.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the position is safe to move to.
+    fn is_safe(&self, x: i32, y: i32) -> bool {
+        x >= 0
+            && y >= 0
+            && x < self.board.len() as i32
+            && y < self.board[0].len() as i32
+            && self.board[x as usize][y as usize] == 0
+    }
+
+    /// Recursively solves the Knight's Tour problem.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: u32) -> bool {
+        if move_count == (self.board.len() * self.board[0].len()) as u32 {
+            return true;
+        }
+        for &(dx, dy) in KnightTour::MOVES.iter() {
+            let next_x = x + dx;
+            let next_y = y + dy;
+
+            if self.is_safe(next_x, next_y) {
+                self.board[next_x as usize][next_y as usize] = (move_count as usize) + 1;
+
+                if self.solve_tour(next_x, next_y, move_count + 1) {
+                    return true;
+                }
+                // Backtrack
+                self.board[next_x as usize][next_y as usize] = 0;
+            }
+        }
+
+        false
+    }
+
+    /// Finds the Knight's Tour starting from the specified position.
+    ///
+    /// # Arguments
+    ///
+    /// * `start_x` - The x-coordinate of the starting position.
+    /// * `start_y` - The y-coordinate of the starting position.
+    ///
+    /// # Returns
+    ///
+    /// A tour matrix if the tour was found or None if not found.
+    fn find_tour(&mut self, start_x: usize, start_y: usize) -> Option<Vec<Vec<usize>>> {
+        if !self.is_safe(start_x as i32, start_y as i32) {
+            return None;
+        }
+
+        self.board[start_x][start_y] = 1;
+
+        if !self.solve_tour(start_x as i32, start_y as i32, 1) {
+            return None;
+        }
+
+        Some(self.board.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! test_find_knight_tour {
+        ($($name:ident: $tc:expr,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                let (size_x, size_y, start_x, start_y, expected) = $tc;
+                assert_eq!(find_knight_tour(size_x, size_y, start_x, start_y), expected);
```

Comment:
What do you think about adding an extra check for the test data?
```suggestion
                if expected.is_some() {
                    assert_eq!(expected.clone().unwrap()[start_x][start_y], 1)
                }
                assert_eq!(find_knight_tour(size_x, size_y, start_x, start_y), expected);
```

Reply:
In my point of view, adding an extra check to verify that the starting position in the expected result is indeed marked as 1 is a good idea. It adds an additional level of validation to ensure that the tour starts from the correct position as specified in the test case. So this can help catch potential issues such as incorrect initialization or indexing errors.